arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/stm/TMVar.kt

```diff
@@ -0,0 +1,38 @@
+package arrow.fx.coroutines.stm
+
+import arrow.fx.coroutines.ConcurrentVar
+import arrow.fx.coroutines.STM
+
+suspend fun <A> STM.newTMVar(a: A): TMVar<A> = TMVar<A>(newTVar(Option.Some(a)))
+suspend fun <A> STM.newEmptyTMVar(): TMVar<A> = TMVar<A>(newTVar(Option.None))
+
+/**
+ * A [TMVar] is the [STM] analog to [ConcurrentVar].
+ * It represents a reference that is either empty or full.
+ *
+ * The main use for [TMVar] is as a synchronization primitive as it can be used to force other transactions
+ *  to wait until a [TMVar] is full.
+ *
+ * ## Creating a [TMVar]:
+ *
+ * As usual with [STM] types there are two equal sets of operators for creating them, one that can be used inside
+ *  and one for use outside of transactions:
+ * - [TMVar.new] and [STM.newTMVar] create a new filled [TMVar]
+ * - [TMVar.empty] and [STM.newEmptyTMVar] create an empty [TMVar]
+ *
+ */
+data class TMVar<A> internal constructor(internal val v: TVar<Option<A>>) {
+  companion object {
+    suspend fun <A> new(a: A): TMVar<A> = TMVar<A>(TVar.new(Option.Some(a)))
+    suspend fun <A> empty(): TMVar<A> = TMVar<A>(TVar.new(Option.None))
+  }
+}
+
+/**
+ * ADT to avoid problems with `TMVar<A?>` because kotlin can't differ `A?` from `A??` so using a nullable
+ *  reference inside `TMVar` does not work.
+ */
+sealed class Option<out A> {
```

Comment:
There is a `Maybe` type somewhere already in Arrow Fx Coroutines. Should we promote this to `typealias Maybe<A> = Either<Unit, A>` and/or move it to the `predef.kt` at the root of the module?

Reply:
Yes in `ConcurrentVar`, I'll switch to using that :+1: We could do `Either<Unit, A>` but that means every `Left(Unit)` is a new object unless we manually create a singleton and use that `const val None = Either.Left(Unit)`.

Reply:
Yeah control flow with nullables only works if you aren't doing it over generics, because that breaks if the generic itself is nullable...

Reply:
Unless you use a nasty hack to optimize 😅 but definitely not worth it until you can benchmark and justify it the maintenance cost.